### DISCUSSION TITLE
Future plans for auto-coding in Fprime

### DISCUSSION THREAD

**Discussion by SMorettini** [nasa/fprime#1965] [Q&A] [answered]
I'm currently working with Fprime and I have a few questions about the future plans for the auto-coding parts of the tool. From my understanding, fpp-to-xml is currently used to generate XML, which is then used to generate hpp/cpp code. However, I also noticed that there is an fpp-to-cpp tool available in the [fpp repository on GitHub](https://github.com/fprime-community/fpp/tree/main/compiler/tools there is fpp-to-cpp).

I'm particularly interested in understanding the direction of the auto-coding parts of Fprime for future development, as I'm considering customizing some of the autogeneration parts. Here are my specific questions:

1. Will future versions of Fprime include a feature that allows fpp-to-cpp to directly produce cpp source from fpp, bypassing the need for XML files?
2. If fpp-to-cpp does generate cpp directly from fpp, will the XML file generation still be necessary, or will it be removed altogether?
3. Will fpp-to-cpp be able to generate cpp code without an Fprime instance ?

Any insights or information about the future direction of Fprime's auto-coding parts would be greatly appreciated, as it will help me make informed decisions about customization options.

Thank you for your time and help.

**Comment by LeStarch** [comment, marked as answer]
The functionality of `fpp-to-cpp` is growing, and once it can perform all of the work that the existing XML autocoders do, `fpp-to-xml` and the `Ai.xml` format will be discontinued.

1. Yes
2. It will be removed eventually
3. Yes, FPP tools do not strictly require F´, however; running them without the aid of F´  means managing a lot more arguments that the build system currently handles

Here is the list of features that need to be added to `fpp-to-cpp` before XML can be entirely done away with. @bocchino feel free to weigh-in if something is wrong here.

1. Topology to CPP
2. Data Types to CPP
3. Ports to CPP
4. Components to CPP
5. Component templates to CPP
6. UTs to CPP
7. FPP to Dictionaries

**Reply by bocchino** [reply]
This looks right. I'll just add the following:
1. 1-3 are done in the devel branch of F Prime. The C++ code for topologies, types, and ports is generated directly from FPP. The XML is still there to support the GDS dictionaries.
2. 4 is in process.
3. The standalone use of FPP tools is described in sections 12-13 of the FPP User's Guide. See https://fprime-community.github.io/fpp/fpp-users-guide.html.

**Reply by SMorettini** [reply]
Do you have a timeline for when the full list of features will be completed and integrated into a Fprime Release?
I'm asking because I'm interested in customizing some parts of Fprime, but I'm unsure whether I should make the changes now or wait for a future release.

**Reply by JackNWhite** [reply]
Is it anticipated that the format or markup language of the dictionaries would change following the deprecation of `*Ai.xml`?

**Reply by LeStarch** [reply]
@bocchino can you fill in any timeline information?

**Reply by bocchino** [reply]
I anticipate having 4 and 5 done this summer. We are still working on the plan for 6 and 7.
